dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java

```diff
@@ -199,10 +199,10 @@ public class WorkflowExecuteRunnable implements Callable<WorkflowSubmitStatus> {
     private final Map<Long, Integer> errorTaskMap = new ConcurrentHashMap<>();
 
     /**
-     * complete task map, taskCode as key, taskInstanceId as value
+     * complete task set
      * in a DAG, only one taskInstance per taskCode is valid
      */
-    private final Map<Long, Integer> completeTaskMap = new ConcurrentHashMap<>();
+    private final Set<Long> completeTaskSet = Sets.newConcurrentHashSet();
```

Comment:
May I ask why to use `set` rather than `map` ?

Reply:
Please see https://github.com/apache/dolphinscheduler/issues/14314

Reply:
This map maybe put some task has code but has no id. And will throw npe, and will retry all the time

Reply:
Yes, I got you. Could we just check not to put these task (which has no id) into the map ? Since other task collections are also `map`

Reply:
WDYT @ruanwenjun

Reply:
Use set is ok here. Additional, this class will be refactor future, use set is a hot fix.